src/Microsoft.OData.Core/MultipartMixed/ODataMultipartMixedBatchReader.cs

```diff
@@ -0,0 +1,472 @@
+                    }
+
+                    // Reset the operation state; the operation state only
+                    // tracks the state of a batch operation while in state Operation.
+                    this.ReaderOperationState = OperationState.None;
+
+                    // Also add a pending ContentId header to the URL resolver now. We ensured above
+                    // that a message has been created for this operation and thus the headers (incl.
+                    // a potential content ID header) have been read.
+                    if (this.ContentIdToAddOnNextRead != null)
+                    {
+                        this.payloadUriConverter.AddContentId(this.ContentIdToAddOnNextRead);
+                        this.ContentIdToAddOnNextRead = null;
+                    }
+
+                    // When we are done with an operation, we have to skip ahead to the next part
+                    // when Read is called again. Note that if the operation stream was never requested
+                    // and the content of the operation has not been read, we'll skip it here.
+                    Debug.Assert(this.ReaderOperationState == OperationState.None, "Operation state must be 'None' at the end of the operation.");
+                    this.State = this.SkipToNextPartAndReadHeaders();
+                    break;
+
+                case ODataBatchReaderState.ChangesetStart:
+                    // When at the start of a changeset, skip ahead to the first operation in the
+                    // changeset (or the end boundary of the changeset).
+                    Debug.Assert(this.batchStream.ChangeSetBoundary != null, "Changeset boundary must have been set by now.");
+                    this.State = this.SkipToNextPartAndReadHeaders();
+                    break;
+
+                case ODataBatchReaderState.ChangesetEnd:
+                    // When at the end of a changeset, reset the changeset boundary and the
+                    // changeset size and then skip to the next part.
+                    this.ResetChangesetSize();
+                    this.batchStream.ResetChangeSetBoundary();
+                    this.State = this.SkipToNextPartAndReadHeaders();
+                    break;
+
+                case ODataBatchReaderState.Exception:    // fall through
+                case ODataBatchReaderState.Completed:
+                    Debug.Assert(false, "Should have checked in VerifyCanRead that we are not in one of these states.");
+                    throw new ODataException(Strings.General_InternalError(InternalErrorCodes.ODataBatchReader_ReadImplementation));
+
+                default:
+                    Debug.Assert(false, "Unsupported reader state " + this.State + " detected.");
+                    throw new ODataException(Strings.General_InternalError(InternalErrorCodes.ODataBatchReader_ReadImplementation));
+            }

```

Comment:
Ideally state transitions would be done in the base class; the derived class would never set the state directly, but would call the next "Startxyz" method in the base class, which would transition state and then call the implementation method in the derived class.  Take a look at how this is done in ODataReader.